// @synfutures/sdks-aggregator/dist/typechain/index.js
"use strict";
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  (function () {
    var ownKeys = function (o) {
      ownKeys =
        Object.getOwnPropertyNames ||
        function (o) {
          var ar = [];
          for (var k in o)
            if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
          return ar;
        };
      return ownKeys(o);
    };
    return function (mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null)
        for (var k = ownKeys(mod), i = 0; i < k.length; i++)
          if (k[i] !== "default") __createBinding(result, mod, k[i]);
      __setModuleDefault(result, mod);
      return result;
    };
  })();
Object.defineProperty(exports, "__esModule", { value: true });
exports.IWETH__factory =
  exports.QuerySplitRoute__factory =
  exports.QuerySingleRoute__factory =
  exports.OysterAggregator__factory =
  exports.Config__factory =
  exports.factories =
    void 0;
exports.factories = __importStar(require("./factories"));
var Config__factory_1 = require("./factories/Config__factory");
Object.defineProperty(exports, "Config__factory", {
  enumerable: true,
  get: function () {
    return Config__factory_1.Config__factory;
  },
});
var OysterAggregator__factory_1 = require("./factories/OysterAggregator__factory");
Object.defineProperty(exports, "OysterAggregator__factory", {
  enumerable: true,
  get: function () {
    return OysterAggregator__factory_1.OysterAggregator__factory;
  },
});
var QuerySingleRoute__factory_1 = require("./factories/QuerySingleRoute__factory");
Object.defineProperty(exports, "QuerySingleRoute__factory", {
  enumerable: true,
  get: function () {
    return QuerySingleRoute__factory_1.QuerySingleRoute__factory;
  },
});
var QuerySplitRoute__factory_1 = require("./factories/QuerySplitRoute__factory");
Object.defineProperty(exports, "QuerySplitRoute__factory", {
  enumerable: true,
  get: function () {
    return QuerySplitRoute__factory_1.QuerySplitRoute__factory;
  },
});
var IWETH__factory_1 = require("./factories/IWETH__factory");
Object.defineProperty(exports, "IWETH__factory", {
  enumerable: true,
  get: function () {
    return IWETH__factory_1.IWETH__factory;
  },
});
