// @derivation-tech/contracts/dist/typechain/index.js
"use strict";
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  (function () {
    var ownKeys = function (o) {
      ownKeys =
        Object.getOwnPropertyNames ||
        function (o) {
          var ar = [];
          for (var k in o)
            if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
          return ar;
        };
      return ownKeys(o);
    };
    return function (mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null)
        for (var k = ownKeys(mod), i = 0; i < k.length; i++)
          if (k[i] !== "default") __createBinding(result, mod, k[i]);
      __setModuleDefault(result, mod);
      return result;
    };
  })();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenAssistant__factory =
  exports.WrappedNative__factory =
  exports.Multicall3__factory =
  exports.ERC721__factory =
  exports.ERC20__factory =
  exports.ERC1155__factory =
  exports.factories =
    void 0;
exports.factories = __importStar(require("./factories"));
var ERC1155__factory_1 = require("./factories/ERC1155__factory");
Object.defineProperty(exports, "ERC1155__factory", {
  enumerable: true,
  get: function () {
    return ERC1155__factory_1.ERC1155__factory;
  },
});
var ERC20__factory_1 = require("./factories/ERC20__factory");
Object.defineProperty(exports, "ERC20__factory", {
  enumerable: true,
  get: function () {
    return ERC20__factory_1.ERC20__factory;
  },
});
var ERC721__factory_1 = require("./factories/ERC721__factory");
Object.defineProperty(exports, "ERC721__factory", {
  enumerable: true,
  get: function () {
    return ERC721__factory_1.ERC721__factory;
  },
});
var Multicall3__factory_1 = require("./factories/Multicall3__factory");
Object.defineProperty(exports, "Multicall3__factory", {
  enumerable: true,
  get: function () {
    return Multicall3__factory_1.Multicall3__factory;
  },
});
var WrappedNative__factory_1 = require("./factories/WrappedNative__factory");
Object.defineProperty(exports, "WrappedNative__factory", {
  enumerable: true,
  get: function () {
    return WrappedNative__factory_1.WrappedNative__factory;
  },
});
var TokenAssistant__factory_1 = require("./factories/TokenAssistant__factory");
Object.defineProperty(exports, "TokenAssistant__factory", {
  enumerable: true,
  get: function () {
    return TokenAssistant__factory_1.TokenAssistant__factory;
  },
});
