// @derivation-tech/context/dist/utils/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TEN_THOUSAND =
  exports.ONE_THOUSAND =
  exports.ONE_HUNDRED =
  exports.MAX_UINT256 =
  exports.TWO =
  exports.ONE =
  exports.ZERO =
  exports.WAD =
  exports.SECS_PER_WEEK =
  exports.SECS_PER_DAY =
  exports.SECS_PER_HOUR =
  exports.SECS_PER_MINUTE =
  exports.MINS_PER_WEEK =
  exports.MINS_PER_DAY =
  exports.MINS_PER_HOUR =
  exports.ZERO_ADDRESS =
  exports.COLON =
  exports.SEMICOLON =
  exports.COMMA =
  exports.HYPHEN =
  exports.ADDR_BATCH_SIZE =
  exports.TOKEN_BATCH_SIZE =
  exports.RIGHT_ARROW =
  exports.PREFIX_SUB =
  exports.PREFIX =
  exports.CHAIN_ID =
    void 0;
const ethers_1 = require("ethers");
var CHAIN_ID;
(function (CHAIN_ID) {
  CHAIN_ID[(CHAIN_ID["LOCAL"] = 31337)] = "LOCAL";
  CHAIN_ID[(CHAIN_ID["ETHEREUM"] = 1)] = "ETHEREUM";
  CHAIN_ID[(CHAIN_ID["ROPSTEN"] = 3)] = "ROPSTEN";
  CHAIN_ID[(CHAIN_ID["RINKEBY"] = 4)] = "RINKEBY";
  CHAIN_ID[(CHAIN_ID["GOERLI"] = 5)] = "GOERLI";
  CHAIN_ID[(CHAIN_ID["KOVAN"] = 42)] = "KOVAN";
  CHAIN_ID[(CHAIN_ID["BSC"] = 56)] = "BSC";
  CHAIN_ID[(CHAIN_ID["POLYGON"] = 137)] = "POLYGON";
  CHAIN_ID[(CHAIN_ID["POLYGONZKEVM"] = 1101)] = "POLYGONZKEVM";
  CHAIN_ID[(CHAIN_ID["ARBITRUM"] = 42161)] = "ARBITRUM";
  CHAIN_ID[(CHAIN_ID["MUMBAI"] = 80001)] = "MUMBAI";
  CHAIN_ID[(CHAIN_ID["ZKSYNC"] = 324)] = "ZKSYNC";
  CHAIN_ID[(CHAIN_ID["LINEA"] = 59144)] = "LINEA";
  CHAIN_ID[(CHAIN_ID["OPTIMISM"] = 10)] = "OPTIMISM";
  CHAIN_ID[(CHAIN_ID["SCROLL"] = 534352)] = "SCROLL";
  CHAIN_ID[(CHAIN_ID["MANTLE"] = 5000)] = "MANTLE";
  CHAIN_ID[(CHAIN_ID["BASE"] = 8453)] = "BASE";
  CHAIN_ID[(CHAIN_ID["CONFLUX"] = 1030)] = "CONFLUX";
  CHAIN_ID[(CHAIN_ID["MAPO"] = 22776)] = "MAPO";
  CHAIN_ID[(CHAIN_ID["SEPOLIA"] = 11155111)] = "SEPOLIA";
  CHAIN_ID[(CHAIN_ID["KLAYTN"] = 8217)] = "KLAYTN";
  CHAIN_ID[(CHAIN_ID["BERABARTIO"] = 80084)] = "BERABARTIO";
  CHAIN_ID[(CHAIN_ID["BLASTSEPOLIA"] = 168587773)] = "BLASTSEPOLIA";
  CHAIN_ID[(CHAIN_ID["BLAST"] = 81457)] = "BLAST";
  CHAIN_ID[(CHAIN_ID["MANTA"] = 169)] = "MANTA";
  CHAIN_ID[(CHAIN_ID["MONADTESTNET"] = 10143)] = "MONADTESTNET";
  CHAIN_ID[(CHAIN_ID["GNOSIS"] = 100)] = "GNOSIS";
  CHAIN_ID[(CHAIN_ID["AVALANCHE"] = 43114)] = "AVALANCHE";
})(CHAIN_ID || (exports.CHAIN_ID = CHAIN_ID = {}));
exports.PREFIX = "├─";
exports.PREFIX_SUB = "├───";
exports.RIGHT_ARROW = "=>";
exports.TOKEN_BATCH_SIZE = 50;
exports.ADDR_BATCH_SIZE = 200;
exports.HYPHEN = "-";
exports.COMMA = ",";
exports.SEMICOLON = ";";
exports.COLON = ":";
exports.ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
exports.MINS_PER_HOUR = 60;
exports.MINS_PER_DAY = 24 * 60;
exports.MINS_PER_WEEK = 7 * 24 * 60;
exports.SECS_PER_MINUTE = 60;
exports.SECS_PER_HOUR = 60 * 60;
exports.SECS_PER_DAY = 24 * 60 * 60;
exports.SECS_PER_WEEK = 7 * 24 * 60 * 60;
exports.WAD = ethers_1.BigNumber.from(String(10 ** 18));
exports.ZERO = ethers_1.BigNumber.from(0);
exports.ONE = ethers_1.BigNumber.from(1);
exports.TWO = ethers_1.BigNumber.from(2);
exports.MAX_UINT256 = ethers_1.BigNumber.from(
  "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
);
exports.ONE_HUNDRED = ethers_1.BigNumber.from(100);
exports.ONE_THOUSAND = ethers_1.BigNumber.from(1000);
exports.TEN_THOUSAND = ethers_1.BigNumber.from(10000);
