// @synfutures/sdks-perp/dist/typechain/index.js
"use strict";
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  (function () {
    var ownKeys = function (o) {
      ownKeys =
        Object.getOwnPropertyNames ||
        function (o) {
          var ar = [];
          for (var k in o)
            if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
          return ar;
        };
      return ownKeys(o);
    };
    return function (mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null)
        for (var k = ownKeys(mod), i = 0; i < k.length; i++)
          if (k[i] !== "default") __createBinding(result, mod, k[i]);
      __setModuleDefault(result, mod);
      return result;
    };
  })();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PythFeederFactory__factory =
  exports.PythFeeder__factory =
  exports.Observer__factory =
  exports.MerkleDistributorWithDeadline__factory =
  exports.Instrument__factory =
  exports.ICexPriceFeeder__factory =
  exports.Guardian__factory =
  exports.Gate__factory =
  exports.EmergingFeederFactory__factory =
  exports.EmergingFeeder__factory =
  exports.DexV2Market__factory =
  exports.Config__factory =
  exports.CexMarket__factory =
  exports.Beacon__factory =
  exports.factories =
    void 0;
exports.factories = __importStar(require("./factories"));
var Beacon__factory_1 = require("./factories/Beacon__factory");
Object.defineProperty(exports, "Beacon__factory", {
  enumerable: true,
  get: function () {
    return Beacon__factory_1.Beacon__factory;
  },
});
var CexMarket__factory_1 = require("./factories/CexMarket__factory");
Object.defineProperty(exports, "CexMarket__factory", {
  enumerable: true,
  get: function () {
    return CexMarket__factory_1.CexMarket__factory;
  },
});
var Config__factory_1 = require("./factories/Config__factory");
Object.defineProperty(exports, "Config__factory", {
  enumerable: true,
  get: function () {
    return Config__factory_1.Config__factory;
  },
});
var DexV2Market__factory_1 = require("./factories/DexV2Market__factory");
Object.defineProperty(exports, "DexV2Market__factory", {
  enumerable: true,
  get: function () {
    return DexV2Market__factory_1.DexV2Market__factory;
  },
});
var EmergingFeeder__factory_1 = require("./factories/EmergingFeeder__factory");
Object.defineProperty(exports, "EmergingFeeder__factory", {
  enumerable: true,
  get: function () {
    return EmergingFeeder__factory_1.EmergingFeeder__factory;
  },
});
var EmergingFeederFactory__factory_1 = require("./factories/EmergingFeederFactory__factory");
Object.defineProperty(exports, "EmergingFeederFactory__factory", {
  enumerable: true,
  get: function () {
    return EmergingFeederFactory__factory_1.EmergingFeederFactory__factory;
  },
});
var Gate__factory_1 = require("./factories/Gate__factory");
Object.defineProperty(exports, "Gate__factory", {
  enumerable: true,
  get: function () {
    return Gate__factory_1.Gate__factory;
  },
});
var Guardian__factory_1 = require("./factories/Guardian__factory");
Object.defineProperty(exports, "Guardian__factory", {
  enumerable: true,
  get: function () {
    return Guardian__factory_1.Guardian__factory;
  },
});
var ICexPriceFeeder__factory_1 = require("./factories/ICexPriceFeeder__factory");
Object.defineProperty(exports, "ICexPriceFeeder__factory", {
  enumerable: true,
  get: function () {
    return ICexPriceFeeder__factory_1.ICexPriceFeeder__factory;
  },
});
var Instrument__factory_1 = require("./factories/Instrument__factory");
Object.defineProperty(exports, "Instrument__factory", {
  enumerable: true,
  get: function () {
    return Instrument__factory_1.Instrument__factory;
  },
});
var MerkleDistributorWithDeadline__factory_1 = require("./factories/MerkleDistributorWithDeadline__factory");
Object.defineProperty(exports, "MerkleDistributorWithDeadline__factory", {
  enumerable: true,
  get: function () {
    return MerkleDistributorWithDeadline__factory_1.MerkleDistributorWithDeadline__factory;
  },
});
var Observer__factory_1 = require("./factories/Observer__factory");
Object.defineProperty(exports, "Observer__factory", {
  enumerable: true,
  get: function () {
    return Observer__factory_1.Observer__factory;
  },
});
var PythFeeder__factory_1 = require("./factories/PythFeeder__factory");
Object.defineProperty(exports, "PythFeeder__factory", {
  enumerable: true,
  get: function () {
    return PythFeeder__factory_1.PythFeeder__factory;
  },
});
var PythFeederFactory__factory_1 = require("./factories/PythFeederFactory__factory");
Object.defineProperty(exports, "PythFeederFactory__factory", {
  enumerable: true,
  get: function () {
    return PythFeederFactory__factory_1.PythFeederFactory__factory;
  },
});
