// @synfutures/sdks-aggregator/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OYSTER_AGGREGATOR_ADDRESS =
  exports.QUERY_SPLIT_ROUTE_ADDRESS =
  exports.QUERY_SINGLE_ROUTE_ADDRESS =
  exports.CONFIG_ADDRESS =
  exports.ZERO_ADDRESS =
  exports.ETH_ADDRESS =
    void 0;
const context_1 = require("@derivation-tech/context");
exports.ETH_ADDRESS = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE";
exports.ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
exports.CONFIG_ADDRESS = {
  [context_1.CHAIN_ID.BASE]: "0xA9A78c647561A3823F1E48b4e151318Ed42C4eC4",
  [context_1.CHAIN_ID.MONADTESTNET]:
    "0x619fae164701e536Be8e1D3A6aaEf35DC7A40fc7",
};
exports.QUERY_SINGLE_ROUTE_ADDRESS = {
  [context_1.CHAIN_ID.BASE]: "0x57C2beA8e020af706f737836061F23ca1E9fd8a5",
  [context_1.CHAIN_ID.MONADTESTNET]:
    "0xB03bB95FAA5DC18D66FAd10A38529f1430bd32e0",
};
exports.QUERY_SPLIT_ROUTE_ADDRESS = {
  [context_1.CHAIN_ID.BASE]: "0x0DE5690A91c559d520A075792DCa1405ebA9ED33",
  [context_1.CHAIN_ID.MONADTESTNET]:
    "0x7a7278dd84B5E63Ada2a4cE3F846b2FF66Fd3cf7",
};
exports.OYSTER_AGGREGATOR_ADDRESS = {
  [context_1.CHAIN_ID.BASE]: "0x58d4adD76f2Ad83C2996f06fcd22b0f15FfBC725",
  [context_1.CHAIN_ID.MONADTESTNET]:
    "0xEf8DD29d887EcD977064Ce169366C95d53926B13",
};
