// @synfutures/sdks-perp/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMPTY_AMM =
  exports.MAX_CANCEL_ORDER_COUNT =
  exports.MIN_BATCH_ORDER_COUNT =
  exports.MAX_BATCH_ORDER_COUNT =
  exports.DEFAULT_REFERRAL_CODE =
  exports.DEFAULT_MAINTENANCE_MARGIN_RATIO =
  exports.DEFAULT_INITIAL_MARGIN_RATIO =
  exports.NATIVE_TOKEN_ADDRESS =
  exports.COMPACT_EMA_PARAM =
  exports.MAINTENANCE_MARGIN_RATIO =
  exports.INITIAL_MARGIN_RATIO =
  exports.PERP_EXPIRY =
  exports.NULL_DDL =
  exports.SETTLING_DURATION =
  exports.MIN_RANGE_MULTIPLIER =
  exports.MIN_ORDER_MULTIPLIER =
  exports.MAX_STABILITY_FEE_RATIO =
  exports.STABILITY_FEE_RATIO_BASE =
  exports.RATIO_BASE =
  exports.RANGE_SPACING =
  exports.ORDER_SPACING =
  exports.PEARL_SPACING =
  exports.INT24_MAX =
  exports.INT24_MIN =
  exports.MIN_TICK =
  exports.MAX_TICK =
  exports.NULL_PRICE =
  exports.ONE_RATIO =
  exports.NULL_RATIO =
  exports.MAX_POSITION_NUM =
  exports.RATIO_DECIMALS =
  exports.WAD_DECIMALS =
    void 0;
const ethers_1 = require("ethers");
exports.WAD_DECIMALS = 18;
exports.RATIO_DECIMALS = 4;
exports.MAX_POSITION_NUM = 128;
exports.NULL_RATIO = 0;
exports.ONE_RATIO = 10000;
exports.NULL_PRICE = ethers_1.BigNumber.from(0);
exports.MAX_TICK = 443636;
exports.MIN_TICK = -322517;
exports.INT24_MIN = -8388608;
exports.INT24_MAX = 8388607;
exports.PEARL_SPACING = 5;
exports.ORDER_SPACING = exports.PEARL_SPACING;
exports.RANGE_SPACING = exports.PEARL_SPACING * 10;
exports.RATIO_BASE = 10000;
exports.STABILITY_FEE_RATIO_BASE = 100;
exports.MAX_STABILITY_FEE_RATIO = Math.pow(2, 16) - 1;
exports.MIN_ORDER_MULTIPLIER = 2;
exports.MIN_RANGE_MULTIPLIER = 10;
exports.SETTLING_DURATION = 30 * 60;
exports.NULL_DDL = Math.pow(2, 32) - 1;
exports.PERP_EXPIRY = exports.NULL_DDL;
exports.INITIAL_MARGIN_RATIO = 1000;
exports.MAINTENANCE_MARGIN_RATIO = 500;
exports.COMPACT_EMA_PARAM = ethers_1.BigNumber.from(
  "1060108913112522979002928172012890034027465765885994002407459484512227426307"
);
exports.NATIVE_TOKEN_ADDRESS = "0x1d6B1d2AD091bec4aAe6A131C92008701531FdaF";
exports.DEFAULT_INITIAL_MARGIN_RATIO = 1000;
exports.DEFAULT_MAINTENANCE_MARGIN_RATIO =
  exports.DEFAULT_INITIAL_MARGIN_RATIO / 2;
exports.DEFAULT_REFERRAL_CODE = "\xff\xff\x00\x00\x00\x00\x00\x00";
exports.MAX_BATCH_ORDER_COUNT = 9;
exports.MIN_BATCH_ORDER_COUNT = 2;
exports.MAX_CANCEL_ORDER_COUNT = 8;
exports.EMPTY_AMM = {
  expiry: 0,
  timestamp: 0,
  status: 0,
  tick: 0,
  sqrtPX96: ethers_1.BigNumber.from(0),
  liquidity: ethers_1.BigNumber.from(0),
  totalLiquidity: ethers_1.BigNumber.from(0),
  involvedFund: ethers_1.BigNumber.from(0),
  openInterests: ethers_1.BigNumber.from(0),
  feeIndex: ethers_1.BigNumber.from(0),
  protocolFee: ethers_1.BigNumber.from(0),
  totalLong: ethers_1.BigNumber.from(0),
  totalShort: ethers_1.BigNumber.from(0),
  longSocialLossIndex: ethers_1.BigNumber.from(0),
  shortSocialLossIndex: ethers_1.BigNumber.from(0),
  longFundingIndex: ethers_1.BigNumber.from(0),
  shortFundingIndex: ethers_1.BigNumber.from(0),
  insuranceFund: ethers_1.BigNumber.from(0),
  settlementPrice: ethers_1.BigNumber.from(0),
  markPrice: ethers_1.BigNumber.from(0),
};
