// @synfutures/sdks-perp/dist/typechain/factories/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PythFeederFactory__factory =
  exports.PythFeeder__factory =
  exports.Observer__factory =
  exports.MerkleDistributorWithDeadline__factory =
  exports.Instrument__factory =
  exports.ICexPriceFeeder__factory =
  exports.Guardian__factory =
  exports.Gate__factory =
  exports.EmergingFeederFactory__factory =
  exports.EmergingFeeder__factory =
  exports.DexV2Market__factory =
  exports.Config__factory =
  exports.CexMarket__factory =
  exports.Beacon__factory =
    void 0;
var Beacon__factory_1 = require("./Beacon__factory");
Object.defineProperty(exports, "Beacon__factory", {
  enumerable: true,
  get: function () {
    return Beacon__factory_1.Beacon__factory;
  },
});
var CexMarket__factory_1 = require("./CexMarket__factory");
Object.defineProperty(exports, "CexMarket__factory", {
  enumerable: true,
  get: function () {
    return CexMarket__factory_1.CexMarket__factory;
  },
});
var Config__factory_1 = require("./Config__factory");
Object.defineProperty(exports, "Config__factory", {
  enumerable: true,
  get: function () {
    return Config__factory_1.Config__factory;
  },
});
var DexV2Market__factory_1 = require("./DexV2Market__factory");
Object.defineProperty(exports, "DexV2Market__factory", {
  enumerable: true,
  get: function () {
    return DexV2Market__factory_1.DexV2Market__factory;
  },
});
var EmergingFeeder__factory_1 = require("./EmergingFeeder__factory");
Object.defineProperty(exports, "EmergingFeeder__factory", {
  enumerable: true,
  get: function () {
    return EmergingFeeder__factory_1.EmergingFeeder__factory;
  },
});
var EmergingFeederFactory__factory_1 = require("./EmergingFeederFactory__factory");
Object.defineProperty(exports, "EmergingFeederFactory__factory", {
  enumerable: true,
  get: function () {
    return EmergingFeederFactory__factory_1.EmergingFeederFactory__factory;
  },
});
var Gate__factory_1 = require("./Gate__factory");
Object.defineProperty(exports, "Gate__factory", {
  enumerable: true,
  get: function () {
    return Gate__factory_1.Gate__factory;
  },
});
var Guardian__factory_1 = require("./Guardian__factory");
Object.defineProperty(exports, "Guardian__factory", {
  enumerable: true,
  get: function () {
    return Guardian__factory_1.Guardian__factory;
  },
});
var ICexPriceFeeder__factory_1 = require("./ICexPriceFeeder__factory");
Object.defineProperty(exports, "ICexPriceFeeder__factory", {
  enumerable: true,
  get: function () {
    return ICexPriceFeeder__factory_1.ICexPriceFeeder__factory;
  },
});
var Instrument__factory_1 = require("./Instrument__factory");
Object.defineProperty(exports, "Instrument__factory", {
  enumerable: true,
  get: function () {
    return Instrument__factory_1.Instrument__factory;
  },
});
var MerkleDistributorWithDeadline__factory_1 = require("./MerkleDistributorWithDeadline__factory");
Object.defineProperty(exports, "MerkleDistributorWithDeadline__factory", {
  enumerable: true,
  get: function () {
    return MerkleDistributorWithDeadline__factory_1.MerkleDistributorWithDeadline__factory;
  },
});
var Observer__factory_1 = require("./Observer__factory");
Object.defineProperty(exports, "Observer__factory", {
  enumerable: true,
  get: function () {
    return Observer__factory_1.Observer__factory;
  },
});
var PythFeeder__factory_1 = require("./PythFeeder__factory");
Object.defineProperty(exports, "PythFeeder__factory", {
  enumerable: true,
  get: function () {
    return PythFeeder__factory_1.PythFeeder__factory;
  },
});
var PythFeederFactory__factory_1 = require("./PythFeederFactory__factory");
Object.defineProperty(exports, "PythFeederFactory__factory", {
  enumerable: true,
  get: function () {
    return PythFeederFactory__factory_1.PythFeederFactory__factory;
  },
});
