// @synfutures/sdks-perp/dist/abis/index.js
"use strict";
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (
          !desc ||
          ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k];
            },
          };
        }
        Object.defineProperty(o, k2, desc);
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
      });
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
      }
    : function (o, v) {
        o["default"] = v;
      });
var __importStar =
  (this && this.__importStar) ||
  (function () {
    var ownKeys = function (o) {
      ownKeys =
        Object.getOwnPropertyNames ||
        function (o) {
          var ar = [];
          for (var k in o)
            if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
          return ar;
        };
      return ownKeys(o);
    };
    return function (mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null)
        for (var k = ownKeys(mod), i = 0; i < k.length; i++)
          if (k[i] !== "default") __createBinding(result, mod, k[i]);
      __setModuleDefault(result, mod);
      return result;
    };
  })();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OBSERVER_ABI =
  exports.INSTRUMENT_ABI =
  exports.GATE_ABI =
  exports.CONFIG_ABI =
  exports.DEXV2_MARKET_ABI =
  exports.CEX_MARKET_ABI =
    void 0;
const CEX_MARKET_ABI = __importStar(require("./CexMarket.json"));
exports.CEX_MARKET_ABI = CEX_MARKET_ABI;
const DEXV2_MARKET_ABI = __importStar(require("./DexV2Market.json"));
exports.DEXV2_MARKET_ABI = DEXV2_MARKET_ABI;
const CONFIG_ABI = __importStar(require("./Config.json"));
exports.CONFIG_ABI = CONFIG_ABI;
const GATE_ABI = __importStar(require("./Gate.json"));
exports.GATE_ABI = GATE_ABI;
const INSTRUMENT_ABI = __importStar(require("./Instrument.json"));
exports.INSTRUMENT_ABI = INSTRUMENT_ABI;
const OBSERVER_ABI = __importStar(require("./Observer.json"));
exports.OBSERVER_ABI = OBSERVER_ABI;
